File: pref/battery/battery_android.java
package org.gioui.x.pref.battery;

import android.os.Build;
import android.content.Context;
import android.os.BatteryManager;
import android.os.PowerManager;
import android.util.Log;

public class battery_android {
    public static int batteryLevel(Context context) {
        if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.LOLLIPOP) {
            return -1;
        }

        BatteryManager bm = (BatteryManager) context.getSystemService(Context.BATTERY_SERVICE);
        if (bm == null) {
            return -1;
        }

        return bm.getIntProperty(BatteryManager.BATTERY_PROPERTY_CAPACITY);
    }

    public static int isCharging(Context context) {
        if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.LOLLIPOP) {
            return -1;
        }

        BatteryManager bm = (BatteryManager) context.getSystemService(Context.BATTERY_SERVICE);
        if (bm == null) {
            return -1;
        }

        if (bm.isCharging()) {
            return 1;
        }
        return 0;
    }

    public static int isSaving(Context context) {
        if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.LOLLIPOP) {
            return -1;
        }

        PowerManager pm = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
        if (pm == null) {
            return -1;
        }

        if (pm.isPowerSaveMode()) {
            return 1;
        }
        return 0;
    }
}

File: pref/battery/battery_android.go
package battery

import (
	"gioui.org/app"
	"gioui.org/x/pref/internal/xjni"
	"git.wow.st/gmp/jni"
)

//go:generate javac -source 8 -target 8 -bootclasspath $ANDROID_HOME/platforms/android-29/android.jar -d $TEMP/pref_battery/classes battery_android.java
//go:generate jar cf battery_android.jar -C $TEMP/pref_battery/classes .

var _Lib = "org/gioui/x/pref/battery/battery_android"

func batteryLevel() (uint8, error) {
	i, err := xjni.DoInt(_Lib, "batteryLevel", "(Landroid/content/Context;)I", jni.Value(app.AppContext()))
	if err != nil || i < 0 {
		return 100, ErrNotAvailableAPI
	}
	return uint8(i), nil
}

func isSavingBattery() (bool, error) {
	i, err := xjni.DoInt(_Lib, "isSaving", "(Landroid/content/Context;)I", jni.Value(app.AppContext()))
	if err != nil || i < 0 {
		return false, ErrNotAvailableAPI
	}
	return i >= 1, err
}

func isCharging() (bool, error) {
	i, err := xjni.DoInt(_Lib, "isCharging", "(Landroid/content/Context;)I", jni.Value(app.AppContext()))
	if err != nil || i < 0 {
		return false, ErrNotAvailableAPI
	}
	return i >= 1, err
}
